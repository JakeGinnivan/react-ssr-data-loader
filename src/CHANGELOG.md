# CHANGELOG

## vNext

- `data-provider.ts`
    - simplify error message as full error details are now provided to consumers