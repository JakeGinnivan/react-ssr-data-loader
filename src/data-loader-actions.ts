export type SuccessAction = {
    type: 'none' | 'initial-fetch' | 'refresh' | 'page' | 'update'
    success: true
}

export type FailedAction = {
    type: 'initial-fetch' | 'refresh' | 'page' | 'update'
    success: false
    error: string
}

export enum LoaderStatus { // The loader is ________ (the data/resource)
    /**
     * The loader is inactive -- not performing any action
     */
    Idle = 'Idle',

    /**
     * The loader has been instantiated and is fetching the resource for the first time
     */
    Fetching = 'Fetching',

    /**
     * The loader is re-fetching the resource
     */
    Refreshing = 'Refreshing',

    /**
     * The loader is fetching the next page of a resurce
     */
    Paging = 'Paging',

    /**
     * The loader is updating the resource
     */
    Updating = 'Updating'
}

export type LoaderState<TData> = {
    status: LoaderStatus
    lastAction: SuccessAction | FailedAction
    /**
     * Some kind of sentinel value so that the object doesn't become top heavy ???
     */

    data: Data<TData>
}

// @ TODO Should we drop dataFromServerSideRender? How do we model not fetching on client
export type Data<TData> =
    | { hasData: true; data: TData; dataFromServerSideRender: boolean }
    | { hasData: false }

export interface DataKeyMap {
    [dataKey: string]: LoaderState<any>
}

export interface DataLoaderState {
    loadingCount: number
    data: { [resourceType: string]: DataKeyMap }
}

export interface ResourceLoadInfo<TAdditionalParameters, TInternalState> {
    resourceType: string
    resourceId: string
    /** Optional additional parameters required to load resource, i.e paging other */
    resourceLoadParams: TAdditionalParameters
    internalState: TInternalState
}

export const INIT = 'resource-data-loader/INIT'
export interface INIT {
    type: typeof INIT
}

export const REFRESH_DATA = 'resource-data-loader/REFRESH_DATA'
// tslint:disable-next-line:class-name
export interface REFRESH_DATA {
    type: typeof REFRESH_DATA
    meta: ResourceLoadInfo<any, any>
}

export const UPDATE_DATA = 'resource-data-loader/UPDATE_DATA'
// tslint:disable-next-line:class-name
export interface UPDATE_DATA {
    type: typeof UPDATE_DATA
    meta: ResourceLoadInfo<any, any>
}

export const NEXT_PAGE = 'resource-data-loader/NEXT_PAGE'
// tslint:disable-next-line:class-name
export interface NEXT_PAGE {
    type: typeof NEXT_PAGE
    meta: ResourceLoadInfo<any, any>
    payload: { existingData: any }
}

export const LOAD_DATA = 'resource-data-loader/LOAD_DATA'
// tslint:disable-next-line:class-name
export interface LOAD_DATA {
    type: 'resource-data-loader/LOAD_DATA'
    meta: ResourceLoadInfo<any, any>
}

export const LOAD_DATA_COMPLETED = 'resource-data-loader/LOAD_DATA_COMPLETED'
// tslint:disable-next-line:class-name
export interface LOAD_DATA_COMPLETED {
    type: 'resource-data-loader/LOAD_DATA_COMPLETED'
    meta: ResourceLoadInfo<any, any>
    payload: {
        data: any
        dataFromServerSideRender: boolean
    }
}

export const LOAD_DATA_FAILED = 'resource-data-loader/LOAD_DATA_FAILED'
// tslint:disable-next-line:class-name
export interface LOAD_DATA_FAILED {
    type: typeof LOAD_DATA_FAILED
    meta: ResourceLoadInfo<any, any>
    payload: Error | string
}

export const UNLOAD_DATA = 'resource-data-loader/UNLOAD_DATA'
// tslint:disable-next-line:class-name
export interface UNLOAD_DATA {
    type: typeof UNLOAD_DATA
    meta: ResourceLoadInfo<any, any>
}

export type Actions =
    | LOAD_DATA
    | LOAD_DATA_COMPLETED
    | LOAD_DATA_FAILED
    | UNLOAD_DATA
    | REFRESH_DATA
    | NEXT_PAGE
    | INIT
    | UPDATE_DATA
