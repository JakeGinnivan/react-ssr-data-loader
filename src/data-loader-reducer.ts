import {
    LoaderState,
    LoaderStatus,
    ResourceLoadInfo,
    DataLoaderState,
    FailedAction,
    Actions,
    LOAD_DATA,
    NEXT_PAGE,
    REFRESH_DATA,
    LOAD_DATA_COMPLETED,
    LOAD_DATA_FAILED,
    UNLOAD_DATA,
    UPDATE_DATA
} from './data-loader-actions'

const defaultState: LoaderState<any> = {
    result: { hasData: false },
    status: LoaderStatus.Idle,
    lastAction: { type: 'none', success: true }
}

const currentDataOrDefault = (
    meta: ResourceLoadInfo<any, any>,
    state: DataLoaderState
): LoaderState<any> => {
    const resourceTypeData = state.data[meta.resourceType]
    if (!resourceTypeData) {
        return defaultState
    }

    const keyData = resourceTypeData[meta.resourceId]
    if (!keyData) {
        return defaultState
    }

    return keyData
}

const statusMap: { [key: string]: FailedAction['type'] } = {
    [LoaderStatus.Paging]: 'page',
    [LoaderStatus.Refreshing]: 'refresh',
    [LoaderStatus.Updating]: 'update',
    [LoaderStatus.Fetching]: 'initial-fetch'
}

export default (
    state: DataLoaderState = {
        data: {},
        loadingCount: 0
    },
    action: Actions
): DataLoaderState => {
    switch (action.type) {
        case LOAD_DATA: {
            const currentOrDefault = currentDataOrDefault(action.meta, state)
            const loading: LoaderState<any> = {
                status: LoaderStatus.Fetching,
                lastAction: currentOrDefault.lastAction, // TODO Should we always go back to idle?
                result: currentOrDefault.result
            }
            return {
                loadingCount: state.loadingCount + 1,
                data: {
                    ...state.data,
                    [action.meta.resourceType]: {
                        ...state.data[action.meta.resourceType],
                        [action.meta.resourceId]: loading
                    }
                }
            }
        }
        // tslint:disable-next-line:no-switch-case-fall-through
        case NEXT_PAGE: {
            const currentState = currentDataOrDefault(action.meta, state)
            const loading: LoaderState<any> = {
                status: LoaderStatus.Paging,
                lastAction: { type: 'none', success: true },
                result: currentState.result
            }
            return {
                loadingCount: state.loadingCount + 1,
                data: {
                    ...state.data,
                    [action.meta.resourceType]: {
                        ...state.data[action.meta.resourceType],
                        [action.meta.resourceId]: loading
                    }
                }
            }
        }
        // tslint:disable-next-line:no-switch-case-fall-through
        case UPDATE_DATA: {
            const currentState = currentDataOrDefault(action.meta, state)
            const loading: LoaderState<any> = {
                status: LoaderStatus.Updating,
                lastAction: { type: 'none', success: true },
                result: currentState.result
            }
            return {
                loadingCount: state.loadingCount + 1,
                data: {
                    ...state.data,
                    [action.meta.resourceType]: {
                        ...state.data[action.meta.resourceType],
                        [action.meta.resourceId]: loading
                    }
                }
            }
        }
        // tslint:disable-next-line:no-switch-case-fall-through
        case REFRESH_DATA: {
            const currentState = currentDataOrDefault(action.meta, state)
            const loading: LoaderState<any> = {
                status: LoaderStatus.Refreshing,
                lastAction: { type: 'none', success: true },
                result: currentState.result
            }
            return {
                loadingCount: state.loadingCount + 1,
                data: {
                    ...state.data,
                    [action.meta.resourceType]: {
                        ...state.data[action.meta.resourceType],
                        [action.meta.resourceId]: loading
                    }
                }
            }
        }
        // tslint:disable-next-line:no-switch-case-fall-through
        case LOAD_DATA_COMPLETED: {
            const currentState = currentDataOrDefault(action.meta, state)
            const lastAction = statusMap[currentState.status]

            const completed: LoaderState<any> = {
                status: LoaderStatus.Idle,
                lastAction: { type: lastAction, success: true },
                result: {
                    hasData: true,
                    data: action.payload.data,
                    dataFromServerSideRender: action.payload.dataFromServerSideRender
                }
            }
            return {
                loadingCount: state.loadingCount - 1,
                data: {
                    ...state.data,
                    [action.meta.resourceType]: {
                        ...state.data[action.meta.resourceType],
                        [action.meta.resourceId]: completed
                    }
                }
            }
        }
        // tslint:disable-next-line:no-switch-case-fall-through
        case LOAD_DATA_FAILED: {
            const currentState = currentDataOrDefault(action.meta, state)
            const lastAction = statusMap[currentState.status]

            const failed: LoaderState<any> = {
                status: LoaderStatus.Idle,
                lastAction: { type: lastAction, success: false, error: action.payload },
                result: currentState.result
            }
            return {
                // this should always occur alongside a LOAD_DATA_COMPLETED so we decrement loadingCount there
                loadingCount: state.loadingCount,
                data: {
                    ...state.data,
                    [action.meta.resourceType]: {
                        ...state.data[action.meta.resourceType],
                        [action.meta.resourceId]: failed
                    }
                }
            }
        }
        // tslint:disable-next-line:no-switch-case-fall-through
        case UNLOAD_DATA: {
            const newState = { loadingCount: state.loadingCount, data: { ...state.data } }
            const dataType = newState.data[action.meta.resourceType]
            delete dataType[action.meta.resourceId]

            if (Object.keys(dataType).length === 0) {
                delete newState.data[action.meta.resourceType]
            } else {
                newState.data[action.meta.resourceType] = dataType
            }

            return newState
        }

        // tslint:disable-next-line:no-switch-case-fall-through
        default:
            return state
    }
}
