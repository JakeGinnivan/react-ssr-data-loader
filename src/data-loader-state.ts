export interface SuccessAction {
    type: 'none' | 'fetch' | 'refresh'
    success: true
}

export interface FailedAction {
    type: 'fetch' | 'refresh'
    success: false
    error: Error & { dataLoadContext?: string }
}

export enum LoaderStatus { // The loader is ________ (the data/resource)
    /**
     * The loader is inactive -- not performing any action
     */
    Idle = 'Idle',

    /**
     * The loader has been instantiated and is fetching the resource for the first time
     */
    Fetching = 'Fetching',

    /**
     * The loader is re-fetching the resource
     */
    Refreshing = 'Refreshing'
}

export interface LoaderState<TData> {
    status: LoaderStatus
    lastAction: SuccessAction | FailedAction
    /**
     * Some kind of sentinel value so that the object doesn't become top heavy ???
     */

    data: Data<TData>
}

// @ TODO Should we drop dataFromServerSideRender? How do we model not fetching on client
export type Data<TData> =
    | { hasData: true; result: TData; dataFromServerSideRender: boolean }
    | { hasData: false }
