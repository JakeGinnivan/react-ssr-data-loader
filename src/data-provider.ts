import * as React from 'react'
import {
    DataLoaderState, LoaderState, Actions, reducer,
    LOAD_DATA, LOAD_DATA_FAILED, LOAD_DATA_COMPLETED,
    UNLOAD_DATA, REFRESH_DATA, NEXT_PAGE,
    INIT, ResourceLoadInfo,
} from './data-loader-actions'
import DataLoaderResources from './data-loader-resources'
import { Subscriptions, DataUpdateCallback } from './subscriptions'

export interface Props {
    initialState?: DataLoaderState
    onError?: (err: string) => void
    loadingCountUpdated?: (loadingCount: number) => void
    loadAllCompleted?: () => void
    stateChanged?: (state: DataLoaderState) => void
    isServerSideRender?: boolean
    resources: DataLoaderResources
}

export interface State extends DataLoaderState {
}

const ssrNeedsData = (state: LoaderState<any> | undefined) => !state || (!state.data.hasData && state.lastAction.success)

export class DataLoaderContext {
    // We need to track this in two places, one with immediate effect,
    // one tied to reacts lifecycle
    private loadingCount = 0
    private state: DataLoaderState
    private subscriptions = new Subscriptions()

    constructor(
        private onStateChanged: (state: DataLoaderState) => void,
        initialState: DataLoaderState | undefined,
        private performLoad: (metadata: ResourceLoadInfo<any>, existingData: any) => Promise<any>,
        private loadAllCompleted: () => void,
        private loadingCountChanged: (loadingCount: number) => void,
        private onError: (err: string) => void,
        public isServerSideRender: boolean
    ) {
        if (initialState) {
            this.state = initialState
        } else {
            this.state = reducer(undefined, { type: INIT })
            onStateChanged(this.state)
        }
    }

    dispatch = <T extends Actions>(action: T): void => {
        this.state = reducer(this.state, action)
        this.subscriptions.notifyStateSubscribersAndDataLoaders(this.state)
        this.onStateChanged(this.state)
    }

    getDataLoaderState = () => this.state

    subscribe = this.subscriptions.subscribeToStateChanges
    unsubscribe = this.subscriptions.unsubscribeFromStateChanges

    async loadData(metadata: ResourceLoadInfo<any>, update: DataUpdateCallback) {
        const firstAttached = this.attach(metadata, update)
        const loadedState = this.getLoadedState(metadata)

        if (this.isServerSideRender && ssrNeedsData(loadedState) && firstAttached) {
            return await this._loadData(metadata)
        } else {
            // Give the data-loader it's state
            const loaderState = this.getLoadedState(metadata)
            loaderState && update(loaderState)
        }

        if (!this.isServerSideRender && firstAttached) {
            // Data is left from a previous session
            if (loadedState && loadedState.data.hasData) {
                if (loadedState.data.dataFromServerSideRender) {
                    // TODO Need to flag data as cached
                }
            } else {
                return await this._loadData(metadata)
            }
        }
    }

    nextPage(metadata: ResourceLoadInfo<any>) {
        const currentState = this.getLoadedState(metadata)
        const existingData = currentState && currentState.data.hasData
            ? currentState.data.data
            : undefined
        this.dispatch<NEXT_PAGE>({
            type: NEXT_PAGE,
            meta: { ...metadata, dataFromServerSideRender: this.isServerSideRender },
            payload: { existingData }
        })

        return this.performLoadData(metadata, existingData)
    }

    refresh(metadata: ResourceLoadInfo<any>) {
        this.dispatch<REFRESH_DATA>({
            type: REFRESH_DATA,
            meta: { ...metadata, dataFromServerSideRender: this.isServerSideRender },
        })

        return this.performLoadData(metadata, undefined)
    }

    unloadData(metadata: ResourceLoadInfo<any>, update: DataUpdateCallback) {
        if (this.detach(metadata, update)) {
            this.dispatch<UNLOAD_DATA>({
                type: UNLOAD_DATA,
                meta: { ...metadata, dataFromServerSideRender: this.isServerSideRender },
            })
        }
    }

    // Returns true when data needs to be unloaded from redux
    detach(metadata: ResourceLoadInfo<any>, update: DataUpdateCallback) {
        const remainingSubscribers = this.subscriptions.unregisterDataLoader(
            metadata.resourceType,
            metadata.resourceId,
            update,
        )

        return remainingSubscribers === 0
    }

    private _loadData = async (metadata: ResourceLoadInfo<any>) => {
        const currentState = this.getLoadedState(metadata)
        const existingData = currentState && currentState.data.hasData
            ? currentState.data.data
            : undefined
        this.dispatch<LOAD_DATA>({
            type: LOAD_DATA,
            meta: { ...metadata, dataFromServerSideRender: this.isServerSideRender }
        })

        await this.performLoadData(metadata, existingData)
    }

    /** @returns true if this is the first data loader to attach to that type and id */
    private attach(metadata: ResourceLoadInfo<any>, update: DataUpdateCallback): boolean {
        const totalDataLoaders = this.subscriptions.registerDataLoader(
            metadata.resourceType,
            metadata.resourceId,
            update,
        )
        const firstDataLoaderForResourceAndId = totalDataLoaders === 1
        return firstDataLoaderForResourceAndId
    }

    private getLoadedState = (metadata: ResourceLoadInfo<any>): LoaderState<any> | undefined => {
        const dataLookup = this.state.data[metadata.resourceType]
        if (!dataLookup) {
            return undefined
        }

        return dataLookup[metadata.resourceId]
    }

    private performLoadData = async (metadata: ResourceLoadInfo<any>, existingData: any) => {
        try {
            this.loadingCount++
            this.loadingCountChanged(this.loadingCount)
            const data = await this.performLoad(metadata, existingData)
            // If we no longer have data loaders, they have been unmounted since we started loading
            if (!this.subscriptions.hasRegisteredDataLoader(metadata.resourceType, metadata.resourceId)) {
                return
            }

            this.dispatch<LOAD_DATA_COMPLETED>({
                type: LOAD_DATA_COMPLETED,
                meta: metadata,
                payload: {
                    data,
                    dataFromServerSideRender: this.isServerSideRender
                }
            })
        } catch (err) {
            // If we no longer have data loaders, they have been unmounted since we started loading
            if (!this.subscriptions.hasRegisteredDataLoader(metadata.resourceType, metadata.resourceId)) {
                return
            }

            let payload: string
            if (err instanceof Error) {
                payload = err.message
            } else {
                payload = err ? err.toString() : ''
            }

            this.onError(`Error when loading ${JSON.stringify(metadata)}:
    ${payload}`)

            this.dispatch<LOAD_DATA_FAILED>({
                type: LOAD_DATA_FAILED,
                meta: { ...metadata, dataFromServerSideRender: this.isServerSideRender },
                payload: payload
            })
        } finally {
            this.loadingCountChanged(this.loadingCount)
            if (--this.loadingCount === 0) {
                this.loadAllCompleted()
            }
        }
    }
}

export default class DataProvider extends React.Component<Props, {}> {
    static childContextTypes = {
        dataLoader: React.PropTypes.object
    }

    private dataLoader: DataLoaderContext
    state: State = reducer(undefined, { type: INIT })

    constructor(props: Props, context: any) {
        super(props, context)

        this.dataLoader = new DataLoaderContext(
            this.props.stateChanged || (() => {}),
            this.props.initialState,
            this.loadData,
            this.props.loadAllCompleted || (() => {}),
            this.props.loadingCountUpdated || (() => {}),
            this.props.onError || (() => {}),
            this.props.isServerSideRender || false,
        )
    }

    private loadData = (metadata: ResourceLoadInfo<any>, existingData: any): Promise<any> => {
        const dataLoader = this.props.resources.getResourceLoader(metadata.resourceType)
        if (!dataLoader) {
            return Promise.reject(`No data loader present for ${metadata.resourceType}`)
        }

        return dataLoader(metadata.resourceId, metadata.resourceLoadParams, existingData)
    }

    getChildContext = (): { dataLoader: DataLoaderContext } => ({ dataLoader: this.dataLoader })

    render() {
        return React.Children.only(this.props.children)
    }
}
