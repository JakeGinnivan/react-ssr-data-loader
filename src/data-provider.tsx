import React from 'react'
import { DataLoaderResources } from './data-loader-resources'
import { DataProviderEvents } from './events'
import { DataLoaderContextComponent } from './data-loader-context'
import {
    DataLoaderStoreAndLoader,
    DataLoaderState,
    ObjectHash,
} from './data-loader-store-and-loader'

export interface Props {
    initialState?: DataLoaderState
    onEvent?: (event: DataProviderEvents) => void | Promise<any>
    isServerSideRender?: boolean
    resources: DataLoaderResources<any>
    globalProps?: object
    /** Override the object hasing function */
    objectHash?: ObjectHash
}

export class DataLoaderProvider extends React.Component<Props> {
    private dataLoader: DataLoaderStoreAndLoader

    constructor(props: Props) {
        super(props)

        this.dataLoader = new DataLoaderStoreAndLoader(
            // tslint:disable-next-line:no-empty
            this.props.onEvent || (() => {}),
            this.props.initialState,
            params => {
                const dataLoader = this.props.resources.getResourceLoader(params.resourceType)
                if (!dataLoader) {
                    return Promise.reject(`No data loader present for ${params.resourceType}`)
                }

                return dataLoader({
                    ...this.props.globalProps,
                    ...params,
                })
            },
            this.props.objectHash || require('hash-sum'),
            this.props.isServerSideRender || false,
        )
    }

    render() {
        return (
            <DataLoaderContextComponent.Provider value={this.dataLoader}>
                {this.props.children}
            </DataLoaderContextComponent.Provider>
        )
    }
}
