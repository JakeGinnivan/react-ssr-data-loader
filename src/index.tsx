export { RenderData, LoadedState as RenderProps } from './data-loader'
export { default as DataProvider } from './data-provider'
export { default as DataLoaderResources } from './data-loader-resources'
export { default as IsLoading } from './is-loading'
