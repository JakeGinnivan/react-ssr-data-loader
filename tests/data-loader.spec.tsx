import React from 'react'
import { ComponentFixture } from './helpers/component-fixture'
import { SharedDataComponentFixture } from './helpers/shared-data-component-fixture'
import { DifferentKeysDataComponentFixture } from './helpers/different-keys-data-component-fixture'
import { DataLoaderProvider } from '../src/data-provider'
import { DataLoaderResources } from '../src/index'
// tslint:disable-next-line:no-implicit-dependencies
import { mount } from 'enzyme'

describe('data-loader', () => {
    it('supports multiple loaders using the same key when data loading', async () => {
        const sut = new SharedDataComponentFixture(undefined, 'testKey', false)

        sut.assertState()
    })

    it('can resolve data from multiple components', async () => {
        const sut = new SharedDataComponentFixture(undefined, 'testKey', false)
        await sut.testDataPromise.resolve({ result: 'Test' })

        sut.assertState()
    })

    it('can load multiple dataloaders with different keys', async () => {
        const sut = new DifferentKeysDataComponentFixture(undefined, 'testKey', 'testKey2', false)
        await sut.testDataPromise.resolve({ result: 'Test' })

        sut.assertState()
    })

    it('multiple components load data once when props change', async () => {
        const sut = new SharedDataComponentFixture(undefined, 'testKey', false)

        await sut.testDataPromise.resolve({ result: 'Success!' })
        sut.resetPromise()
        sut.root.setProps({ resourceId: 'newData' })
        await sut.testDataPromise.resolve({ result: 'Success2!' })

        sut.assertState()
    })

    it('data is not unloaded until the last attached data-loader is unmounted', async () => {
        const sut = new SharedDataComponentFixture(undefined, 'testKey', false, false, true)

        await sut.testDataPromise.resolve({ result: 'Success!' })
        sut.root.unmount()
        await new Promise(resolve => setTimeout(resolve))

        sut.assertState()
    })

    it('handles onEvent throwing gracefully', async () => {
        const errors: string[] = []
        // tslint:disable-next-line:no-console
        console.error = jest.fn((...err: any[]) => {
            errors.push(JSON.stringify(err))
        })
        const sut = new ComponentFixture(undefined, 'testKey', {
            isServerSideRender: false,
            onEvent: () => {
                throw new Error('Boom')
            }
        })
        await sut.testDataPromise.resolve({ result: 'Test' })

        sut.assertState()
        expect(errors).toMatchSnapshot()
    })

    it('ignores completion if unmounted first', async () => {
        const sut = new ComponentFixture(undefined, 'testKey', { isServerSideRender: false })
        await sut.unmount()
        await sut.testDataPromise.resolve({ result: 'Test' })

        sut.assertState()
    })

    it('notifies when all work is completed', async () => {
        const sut = new ComponentFixture(undefined, 'testKey', { isServerSideRender: false })

        expect(sut.loadAllCompletedCalled).toBe(0)
        await sut.testDataPromise.resolve({ result: 'Test' })
        expect(sut.loadAllCompletedCalled).toBe(1)
    })

    it('can support preserving data on unmount', async () => {
        let sut = new ComponentFixture(undefined, 'testKey', {
            isServerSideRender: false,
            unloadDataOnUnmount: false
        })
        await sut.testDataPromise.resolve({ result: 'Test' })
        await sut.unmount()

        sut.assertState()
        // Check we can re-mount the existing data
        sut = new ComponentFixture(sut.currentState, 'testKey', {
            isServerSideRender: false,
            unloadDataOnUnmount: false
        })
        sut.assertState()
    })

    it('throws when the same resource is registered multiple times', () => {
        const resources = new DataLoaderResources()

        resources.registerResource('test', () => Promise.resolve('test'))
        expect(() => resources.registerResource('test', () => Promise.resolve('test'))).toThrow()
    })

    it('throws when the same paged resource is registered multiple times', () => {
        const resources = new DataLoaderResources()

        resources.registerResource('test', () => Promise.resolve('test'))
        expect(() => resources.registerResource('test', () => Promise.resolve('test'))).toThrow()
    })

    it('resource can resolve synchronously', async () => {
        const resources = new DataLoaderResources()

        const LoadTest = resources.registerResource('test', () => 'Result!')
        let loadCount = 0

        const wrapper = mount(
            <DataLoaderProvider resources={resources}>
                <LoadTest
                    resourceId="Test!"
                    // tslint:disable-next-line:jsx-no-lambda
                    renderData={renderProps => {
                        loadCount++
                        if (renderProps.data.hasData) {
                            return <div>{renderProps.data.result}</div>
                        }

                        return <div>No data!</div>
                    }}
                />
            </DataLoaderProvider>
        )

        expect(wrapper.html()).toMatchSnapshot()
        await new Promise(resolve => setTimeout(resolve))

        expect(loadCount).toBe(1)
    })

    it('data loader handles synchronous throw', async () => {
        const resources = new DataLoaderResources()

        const LoadTest = resources.registerResource('test', () => {
            throw new Error('Synchronous fail')
        })
        let loadCount = 0

        const wrapper = mount(
            <DataLoaderProvider resources={resources}>
                <LoadTest
                    resourceId="Test!"
                    // tslint:disable-next-line:jsx-no-lambda
                    renderData={renderProps => {
                        loadCount++
                        if (!renderProps.lastAction.success) {
                            return <div>{renderProps.lastAction.error.message}</div>
                        }
                        if (renderProps.data.hasData) {
                            return <div>{renderProps.data.result}</div>
                        }

                        return <div>No data!</div>
                    }}
                />
            </DataLoaderProvider>
        )

        expect(wrapper.html()).toMatchSnapshot()
        await new Promise(resolve => setTimeout(resolve))

        expect(loadCount).toBe(1)
    })

    it('resource can resolve synchronously when resource changes', async () => {
        const sut = new ComponentFixture(undefined, 'testKey', {
            isServerSideRender: false,
            unloadDataOnUnmount: false,
            syncResult: {
                result: 'Result!'
            }
        })

        sut.testDataResult = {
            result: 'Result2!'
        }
        sut.root.setProps({ resourceId: 'newData' })

        sut.assertState()
    })
})
