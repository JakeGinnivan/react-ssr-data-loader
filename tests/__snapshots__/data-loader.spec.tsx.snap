exports[`data-loader can load multiple dataloaders with different keys 1`] = `
Object {
  "loadAllCompletedCalled": 0,
  "loadDataCount": 2,
  "renderCount": 2,
  "renderProps": Object {
    "data": Object {
      "data": Object {
        "result": "Test",
      },
      "dataFromServerSideRender": false,
      "hasData": true,
    },
    "lastAction": Object {
      "success": true,
      "type": "initial-fetch",
    },
    "status": 0,
  },
  "renderProps2": Object {
    "data": Object {
      "hasData": false,
    },
    "lastAction": Object {
      "success": true,
      "type": "none",
    },
    "status": 1,
  },
}
`;

exports[`data-loader can resolve data from multiple components 1`] = `
Object {
  "loadAllCompletedCalled": 1,
  "loadDataCount": 1,
  "renderCount": 2,
  "renderCount2": 2,
  "renderProps1": Object {
    "data": Object {
      "data": Object {
        "result": "Test",
      },
      "dataFromServerSideRender": false,
      "hasData": true,
    },
    "lastAction": Object {
      "success": true,
      "type": "initial-fetch",
    },
    "status": 0,
  },
  "renderProps2": Object {
    "data": Object {
      "data": Object {
        "result": "Test",
      },
      "dataFromServerSideRender": false,
      "hasData": true,
    },
    "lastAction": Object {
      "success": true,
      "type": "initial-fetch",
    },
    "status": 0,
  },
}
`;

exports[`data-loader can specify arguments for data loader 1`] = `
Object {
  "loadAllCompletedCalled": 1,
  "loadDataCount": 1,
  "renderActions": Object {
    "refresh": [Function],
  },
  "renderCount": 2,
  "renderProps": Object {
    "data": Object {
      "data": Object {
        "result": "Test",
      },
      "dataFromServerSideRender": false,
      "hasData": true,
    },
    "lastAction": Object {
      "success": true,
      "type": "initial-fetch",
    },
    "status": 0,
  },
}
`;

exports[`data-loader can support preserving data on unmount 1`] = `
Object {
  "loadAllCompletedCalled": 1,
  "loadDataCount": 1,
  "renderActions": Object {
    "refresh": [Function],
  },
  "renderCount": 2,
  "renderProps": Object {
    "data": Object {
      "data": Object {
        "result": "Test",
      },
      "dataFromServerSideRender": false,
      "hasData": true,
    },
    "lastAction": Object {
      "success": true,
      "type": "initial-fetch",
    },
    "status": 0,
  },
}
`;

exports[`data-loader can support preserving data on unmount 2`] = `
Object {
  "loadAllCompletedCalled": 0,
  "loadDataCount": 0,
  "renderActions": Object {
    "refresh": [Function],
  },
  "renderCount": 1,
  "renderProps": Object {
    "data": Object {
      "data": Object {
        "result": "Test",
      },
      "dataFromServerSideRender": false,
      "hasData": true,
    },
    "lastAction": Object {
      "success": true,
      "type": "initial-fetch",
    },
    "status": 0,
  },
}
`;

exports[`data-loader ignores completion if unmounted first 1`] = `
Object {
  "loadAllCompletedCalled": 1,
  "loadDataCount": 1,
  "renderActions": Object {
    "refresh": [Function],
  },
  "renderCount": 1,
  "renderProps": Object {
    "data": Object {
      "hasData": false,
    },
    "lastAction": Object {
      "success": true,
      "type": "none",
    },
    "status": 1,
  },
}
`;

exports[`data-loader multiple components load data once when props change 1`] = `
Object {
  "loadAllCompletedCalled": 2,
  "loadDataCount": 2,
  "renderCount": 4,
  "renderCount2": 4,
  "renderProps1": Object {
    "data": Object {
      "data": Object {
        "result": "Success2!",
      },
      "dataFromServerSideRender": false,
      "hasData": true,
    },
    "lastAction": Object {
      "success": true,
      "type": "initial-fetch",
    },
    "status": 0,
  },
  "renderProps2": Object {
    "data": Object {
      "data": Object {
        "result": "Success2!",
      },
      "dataFromServerSideRender": false,
      "hasData": true,
    },
    "lastAction": Object {
      "success": true,
      "type": "initial-fetch",
    },
    "status": 0,
  },
}
`;

exports[`data-loader supports multiple loaders using the same key when data loading 1`] = `
Object {
  "loadAllCompletedCalled": 0,
  "loadDataCount": 1,
  "renderCount": 1,
  "renderCount2": 1,
  "renderProps1": Object {
    "data": Object {
      "hasData": false,
    },
    "lastAction": Object {
      "success": true,
      "type": "none",
    },
    "status": 1,
  },
  "renderProps2": Object {
    "data": Object {
      "hasData": false,
    },
    "lastAction": Object {
      "success": true,
      "type": "none",
    },
    "status": 1,
  },
}
`;

exports[`data-loader supports paged data 1`] = `
Object {
  "loadAllCompletedCalled": 1,
  "loadDataCount": 1,
  "renderActions": Object {
    "nextPage": [Function],
    "refresh": [Function],
  },
  "renderCount": 2,
  "renderProps": Object {
    "data": Object {
      "data": Object {
        "data": Array [
          "Test",
        ],
        "pageNumber": 1,
      },
      "dataFromServerSideRender": false,
      "hasData": true,
    },
    "lastAction": Object {
      "success": true,
      "type": "initial-fetch",
    },
    "status": 0,
  },
}
`;

exports[`data-loader supports paged data 2`] = `
Object {
  "loadAllCompletedCalled": 1,
  "loadDataCount": 2,
  "renderActions": Object {
    "nextPage": [Function],
    "refresh": [Function],
  },
  "renderCount": 3,
  "renderProps": Object {
    "data": Object {
      "data": Object {
        "data": Array [
          "Test",
        ],
        "pageNumber": 1,
      },
      "dataFromServerSideRender": false,
      "hasData": true,
    },
    "lastAction": Object {
      "success": true,
      "type": "none",
    },
    "status": 3,
  },
}
`;

exports[`data-loader supports paged data 3`] = `
Object {
  "loadAllCompletedCalled": 2,
  "loadDataCount": 2,
  "renderActions": Object {
    "nextPage": [Function],
    "refresh": [Function],
  },
  "renderCount": 4,
  "renderProps": Object {
    "data": Object {
      "data": Object {
        "data": Array [
          "Test",
          "Test2",
        ],
        "pageNumber": 2,
      },
      "dataFromServerSideRender": false,
      "hasData": true,
    },
    "lastAction": Object {
      "success": true,
      "type": "page",
    },
    "status": 0,
  },
}
`;
